File: eclipse-initial/src/native/org.eclipse.ice.reactor.native/Tube.cpp
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
#include "Tube.h"
#include <H5Cpp.h>
#include <HdfWriterFactory.h>
#include <HdfReaderFactory.h>
#include <iostream>
#include <memory>
#include "UtilityOperations.h"
#include <IHdfReadable.h>
#include <IHdfWriteable.h>
#include "Material.h"
#include "HDF5LWRTagType.h"
#include "TubeType.h"

using namespace ICE_Reactor;

Tube::Tube(Tube & arg) : Ring(arg) {

    //begin-user-code

    //Copy contents
    this->tubeType = arg.tubeType;

    //end-user-code

}
Tube::~Tube() {
    //TODO Auto-generated method stub
}
Tube::Tube() {

    //begin-user-code

    //Setup Defaults
    this->name = "Tube";
    this->id= 1;
    this->description= "Tube's Description";
    this->innerRadius = 0;
    this->outerRadius = 1;
    this->height = 1;
    this->tubeType = GUIDE;
    this->material = std::shared_ptr<Material> (new Material());
    this->HDF5LWRTag = TUBE;

    //end-user-code

}
Tube::Tube(const std::string name) {

    //begin-user-code

    //Setup Defaults
    this->name = "Tube";
    this->id= 1;
    this->description= "Tube's Description";
    this->innerRadius = 0;
    this->outerRadius = 1;
    this->height = 1;
    this->tubeType = GUIDE;
    this->material = std::shared_ptr<Material> (new Material());
    this->HDF5LWRTag = TUBE;

    //Run setters
    Tube::setName(name);

    //end-user-code

}
Tube::Tube(const std::string name, TubeType tubeType) {
    //begin-user-code

    //Setup Defaults
    this->name = "Tube";
    this->id= 1;
    this->description= "Tube's Description";
    this->innerRadius = 0;
    this->outerRadius = 1;
    this->height = 1;
    this->tubeType = GUIDE;
    this->material = std::shared_ptr<Material> (new Material());
    this->HDF5LWRTag = TUBE;


    //Run setters
    Tube::setName(name);
    Tube::setTubeType(tubeType);

    //end-user-code
}
Tube::Tube(const std::string name, TubeType tubeType, std::shared_ptr<Material> material, double height, double outerRadius) {
    //begin-user-code

    //Setup Defaults
    this->name = "Tube";
    this->id= 1;
    this->description= "Tube's Description";
    this->innerRadius = 0;
    this->outerRadius = 1;
    this->height = 1;
    this->tubeType = GUIDE;
    this->material = std::shared_ptr<Material> (new Material());
    this->HDF5LWRTag = TUBE;


    //Run setters
    Tube::setName(name);
    Tube::setTubeType(tubeType);
    Tube::setMaterial(material);
    Tube::setHeight(height);
    Tube::setOuterRadius(outerRadius);

    //end-user-code
}
Tube::Tube(const std::string name, TubeType tubeType, std::shared_ptr<Material> material, double height, double innerRadius, double outerRadius) {
    //begin-user-code

    //Setup Defaults
    this->name = "Tube";
    this->id= 1;
    this->description= "Tube's Description";
    this->innerRadius = 0;
    this->outerRadius = 1;
    this->height = 1;
    this->tubeType = GUIDE;
    this->material = std::shared_ptr<Material> (new Material());
    this->HDF5LWRTag = TUBE;


    //Run setters
    Tube::setName(name);
    Tube::setTubeType(tubeType);
    Tube::setMaterial(material);
    Tube::setHeight(height);
    Tube::setOuterRadius(outerRadius);
    Tube::setInnerRadius(innerRadius);

    //end-user-code
}
TubeType Tube::getTubeType() {

    // begin-user-code

    return this->tubeType;

    // end-user-code
}
void Tube::setTubeType(TubeType tubeType) {

    //begin-user-code

    this->tubeType = tubeType;


    return;
    // end-user-code

}
bool Tube::operator==(const Tube& otherObject) const {

    // begin-user-code

    return Ring::operator ==(otherObject) && this->tubeType == otherObject.tubeType;

    // end-user-code
}

std::shared_ptr<ICE_DS::Identifiable> Tube::clone() {
    // begin-user-code

    // Local Declarations
    std::shared_ptr<Tube> component(new Tube (*this));

    // Return the newly instantiated object
    return component;

    // end-user-code
}
bool Tube::writeAttributes(std::shared_ptr<H5::H5File> h5File, std::shared_ptr<H5::Group> h5Group) {

    // begin-user-code

    bool flag = true;

    flag &= Ring::writeAttributes(h5File, h5Group);
    flag &= ICE_IO::HdfWriterFactory::writeStringAttribute(h5File, h5Group, "tubeType", UtilityOperations::toStringTubeType(this->tubeType));

    return flag;


    // end-user-code
}
bool Tube::readAttributes(std::shared_ptr<H5::Group> h5Group) {

    // begin-user-code

    //Local Declarations
    TubeType type;

    //If group is null or the super read attributes is false, return false
    if(h5Group.get() == NULL || Ring::readAttributes(h5Group) == false) return false;

    //Try to get values.  If failure, return false;
    try {
        type  = UtilityOperations::fromStringTubeType(ICE_IO::HdfReaderFactory::readStringAttribute(h5Group, "tubeType"));
    } catch (...) {
        return false;
    }

    //Setup the values
    this->tubeType = type;

    //Operation passed!
    return true;

    // end-user-code

}

File: src/org.eclipse.ice.reactor/src/org/eclipse/ice/reactor/Tube.java
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor;

import ncsa.hdf.object.h5.H5File;
import ncsa.hdf.object.h5.H5Group;

import org.eclipse.ice.io.hdf.HdfReaderFactory;
import org.eclipse.ice.io.hdf.HdfWriterFactory;

/**
 * <p>
 * The Tube class represents the hollow tubes in a FuelAssembly which allow for
 * the insertion of discrete poison rodlets (Guide Tubes) and instrument
 * thimbles (Instrument Tube).
 * </p>
 * 
 * @author Scott Forest Hull II
 */
public class Tube extends Ring {
	/**
	 * <p>
	 * One of the TubeType enumeration values.
	 * </p>
	 * 
	 */
	private TubeType tubeType;

	/**
	 * <p>
	 * The nullary Constructor.
	 * </p>
	 * 
	 */
	public Tube() {
		// Call super constructor - sets up defaults
		super();

		// Set name and description for tube
		this.name = "Tube";
		this.description = "Tube's Description";

		// Setup other attributes for tube
		this.tubeType = TubeType.GUIDE;

		// Setup the LWRComponentType to the correct type
		this.HDF5LWRTag = HDF5LWRTagType.TUBE;

	}

	/**
	 * <p>
	 * The parameterized Constructor.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The name of this Tube.
	 *            </p>
	 */
	public Tube(String name) {
		// Call nullary constructor
		this();

		// Set the name
		this.setName(name);

	}

	/**
	 * <p>
	 * A parameterized Constructor.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The name of this Tube.
	 *            </p>
	 * @param tubeType
	 *            <p>
	 *            The TubeType enumeration value for this Tube.
	 *            </p>
	 */
	public Tube(String name, TubeType tubeType) {

		// Call non-nullary constructor
		this(name);

		// Call extra operations
		this.setTubeType(tubeType);

	}

	/**
	 * <p>
	 * A parameterized Constructor.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The name of this Tube.
	 *            </p>
	 * @param tubeType
	 *            <p>
	 *            The TubeType enumeration value for this Tube.
	 *            </p>
	 * @param material
	 *            <p>
	 *            The Material for this Tube.
	 *            </p>
	 * @param height
	 *            <p>
	 *            The height of this Tube, which must be greater than zero.
	 *            </p>
	 * @param outerRadius
	 *            <p>
	 *            The outer radius of this Tube, which must be greater than the
	 *            innerRadius value.
	 *            </p>
	 */
	public Tube(String name, TubeType tubeType, Material material,
			double height, double outerRadius) {
		// Call non-nullary constructor
		this(name, tubeType);

		// Set other values
		super.setMaterial(material);
		super.setHeight(height);
		super.setOuterRadius(outerRadius);

	}

	/**
	 * <p>
	 * A parameterized Constructor.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The name of this Tube.
	 *            </p>
	 * @param tubeType
	 *            <p>
	 *            The TubeType enumeration value for this Tube.
	 *            </p>
	 * @param material
	 *            <p>
	 *            The Material for this Tube.
	 *            </p>
	 * @param height
	 *            <p>
	 *            The height of this Tube, which must be greater than zero.
	 *            </p>
	 * @param innerRadius
	 *            <p>
	 *            The inner radius of this Tube, which must be greater than or
	 *            equal to zero.
	 *            </p>
	 * @param outerRadius
	 *            <p>
	 *            The outer radius of this Tube, which must be greater than the
	 *            innerRadius value.
	 *            </p>
	 */
	public Tube(String name, TubeType tubeType, Material material,
			double height, double innerRadius, double outerRadius) {
		// Call non-nullary constructor
		this(name, tubeType, material, height, outerRadius);

		// Set other values
		this.setInnerRadius(innerRadius);

	}

	/**
	 * <p>
	 * Returns the TubeType enumeration value for this Tube.
	 * </p>
	 * 
	 * @return <p>
	 *         The TubeType enumeration value for this Tube.
	 *         </p>
	 */
	public TubeType getTubeType() {

		return this.tubeType;
	}

	/**
	 * <p>
	 * Sets the TubeType enumeration value for this Tube. Can not set to null.
	 * </p>
	 * 
	 * @param tubeType
	 *            <p>
	 *            The TubeType enumeration value for this Tube.
	 *            </p>
	 */
	public void setTubeType(TubeType tubeType) {

		// If tube type is not null
		if (tubeType != null) {
			this.tubeType = tubeType;
		}

	}

	/**
	 * <p>
	 * Overrides the equals operation to check the attributes on this object
	 * with another object of the same type. Returns true if the objects are
	 * equal. False otherwise.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be compared.
	 *            </p>
	 * @return <p>
	 *         True if otherObject is equal. False otherwise.
	 *         </p>
	 */
	@Override
	public boolean equals(Object otherObject) {

		// Local Declarations
		Tube tube;
		boolean retVal = false;

		// If they are equal to the same object, return true
		if (this == otherObject) {
			return true;
		}
		if (otherObject != null && otherObject instanceof Tube) {

			// Cast
			tube = (Tube) otherObject;

			// Check values
			retVal = super.equals(otherObject)
					&& this.tubeType.equals(tube.tubeType);
		}

		// Return retVal
		return retVal;

	}

	/**
	 * <p>
	 * Returns the hashCode of the object.
	 * </p>
	 * 
	 * @return <p>
	 *         The hash of the object.
	 *         </p>
	 */
	@Override
	public int hashCode() {

		// Get hash from super
		int hash = super.hashCode();

		// Get local hash
		hash += 31 * this.tubeType.hashCode();

		// Return the hashCode
		return hash;

	}

	/**
	 * <p>
	 * Deep copies the contents of the object.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be copied.
	 *            </p>
	 */
	public void copy(Tube otherObject) {

		// If the otherObject is null, return
		if (otherObject == null) {
			return;
		}
		// Copy contents - super
		super.copy(otherObject);

		// copy local contents
		this.tubeType = otherObject.tubeType;

	}

	/**
	 * <p>
	 * Deep copies and returns a newly instantiated object.
	 * </p>
	 * 
	 * @return <p>
	 *         The newly instantiated copied object.
	 *         </p>
	 */
	@Override
	public Object clone() {
		// Local Declarations
		Tube tube = new Tube();

		// Copy contents
		tube.copy(this);

		// Return newly instantiated object
		return tube;

	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public boolean writeAttributes(H5File h5File, H5Group h5Group) {
		boolean flag = true;

		flag &= super.writeAttributes(h5File, h5Group);
		flag &= HdfWriterFactory.writeStringAttribute(h5File, h5Group,
				"tubeType", tubeType.toString());

		return flag;
	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public boolean readAttributes(H5Group h5Group) {

		// Local Declarations
		boolean flag = true;

		// Get values
		TubeType type1 = TubeType.GUIDE;
		TubeType type = type1.toType((HdfReaderFactory.readStringAttribute(
				h5Group, "tubeType")));

		// Call super
		flag &= super.readAttributes(h5Group);

		// check values
		if (type == null || !flag || h5Group == null) {
			return false;
		}
		// If everything is valid, then set data
		this.tubeType = type;

		return true;

	}

	/**
	 * <p>
	 * This operation accepts an ILWRComponentVisitor that can be visit the
	 * LWRComponent to ascertain its type and perform various type-specific
	 * operations.
	 * </p>
	 * 
	 * @param visitor
	 *            <p>
	 *            The visitor
	 *            </p>
	 */
	@Override
	public void accept(ILWRComponentVisitor visitor) {
		visitor.visit(this);
	}
}